File: DP_Fibonacci.py
#Dynamic Programming 피보나치 수열 비교

#Memoiztion : Top-Down 가끔 재귀 횟수 제한 오류가 걸릴 수도 있음
'''
import time

d = [0] * 50

def fibo(x):
    if x == 1 or x == 2:
        return 1
    if d[x] != 0:
        return d[x]
    d[x] = fibo(x-1) + fibo(x-2)
    return d[x]

for num in range(5, 6):
    start = time.time()
    res = fibo(num)
    print(res, '-> 러닝타임:', round(time.time() - start, 2), '초')
'''

#DP Table : Bottom-Up
'''
d = [0] * 100
d[1] = 1
d[2] = 1
N = 99 #피보나치 수열의 99번째 숫자는?

for i in range(3, N+1):
    d[i] = d[i-1] + d[i-2]

print(d[N])

#출처 https://techblog-history-younghunjo1.tistory.com/183

File: Algorithm/DP_Fibonacci.c
#include <stdio.h>

/*Memoiztion : Top-Down 가끔 재귀 횟수 제한 오류가 걸릴 수도 있음
int fibo(int n){
    int a;
    int d[30] = {0};
    if((n==1)||(n==2)) return 1;
    if(d[n] != 0){
        return d[n];
    }
    d[n] = fibo(n-1) + fibo(n-2);
    return d[n];
}

int main(void){
    int x, result;
    scanf("%d", &x);
    result = fibo(x);
    printf("%d", result);
}*/


/*Bottom-Up
int fibo(int n){
    int a;
    int d[30];
    d[0] = 0;
    d[1] = 1;

    for (a = 2; a<=n; a++){
        d[a] = d[a-2] + d[a-1];
    }
    return d[n];
}

int main(){
    int x, result;
    scanf("%d", &x);
    result = fibo(x);
    printf("%d", result);
}*/
